File: multiplication_table.cpp
#include <iostream>
#include <iomanip>
#include <string>

using namespace std;

int main(int argc, char* argv[]) {
    setlocale(LC_ALL, "ru_RU.UTF-8");
    int n, m;
    if (argc >= 3) {
        try {
            n = stoi(argv[1]);
            m = stoi(argv[2]);
        } catch (const invalid_argument&) {
            cerr << "Ошибка: введите целые числа." << endl;
            return 1;
        }
    } else {
        cout << "Введите количество строк (N): ";
        cin >> n;
        cout << "Введите количество столбцов (M): ";
        cin >> m;
        
        if (cin.fail()) {
            cerr << "Ошибка: введите целые числа." << endl;
            return 1;
        }
    }

    if (n <= 0 || m <= 0) {
        cerr << "Размеры таблицы должны быть положительными." << endl;
        return 1;
    }

    int max_val = n * m;
    int col_width = static_cast<int>(to_string(max_val).length()) + 1;

    cout << "\nТаблица умножения:" << endl;

    // Верхний заголовок
    cout << setw(col_width) << " ";
    for (int j = 1; j <= m; ++j) {
        cout << setw(col_width) << j;
    }
    cout << endl;

    // Разделитель
    cout << string(col_width * (m + 1), '-') << endl;

    // Тело таблицы
    for (int i = 1; i <= n; ++i) {
        cout << setw(col_width - 1) << left << i << "|";
        for (int j = 1; j <= m; ++j) {
            cout << setw(col_width) << right << i * j;
        }
        cout << endl;
    }

    return 0;
}

File: multiplication_table.py
import sys

def main():
    # Проверка аргументов командной строки (для совместимости с C++)
    if len(sys.argv) > 1:
        try:
            n = int(sys.argv[1])
            m = int(sys.argv[2])
        except ValueError:
            print("Ошибка: введите целые числа.")
            return
    else:
        try:
            n = int(input("Введите количество строк (N): "))
            m = int(input("Введите количество столбцов (M): "))
        except ValueError:
            print("Ошибка: введите целые числа.")
            return

    if n <= 0 or m <= 0:
        print("Размеры таблицы должны быть положительными.")
        return

    # Определяем ширину столбца для выравнивания
    # Самое большое число n * m  но столбцов M поэтому смотрим на max(n*i)
    max_val = n * m
    col_width = len(str(max_val)) + 1  # +1 для отступа

    print("\nТаблица умножения:")
    # Заголовок столбцов
    print(" " * (col_width - 1), end="")
    for j in range(1, m + 1):
        print(f"{j:>{col_width}}", end="")
    print()
    
    # Разделитель
    print("-" * (col_width * (m + 1)))

    for i in range(1, n + 1):
        # Заголовок строки
        print(f"{i:<{col_width - 1}}|", end="")
        for j in range(1, m + 1):
            print(f"{i * j:{col_width}}", end="")
        print()

if __name__ == "__main__":
    main()